---
title: "Quality control using Plink"
author: "Maria Izabel Cavassim Alves"
date: "11/24/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Helpful links if you get stuck during this exercise:

* [PLINK2 documentation](https://www.cog-genomics.org/plink2/)
* [Introduction to the linux command line](http://lifehacker.com/5633909/who-needs-a-mouse-learn-to-use-the-commandline-for-almost-anything)


# Accessing hoffman2
If you have a macbook or linux, you can open the terminal and type the following:
```{bash, eval=F}
ssh [YOUR_LOGIN_NAME]@hoffman2.idre.ucla.edu
```
It will request your password and once you type it you are in!Yay!

If you have a windows machine, then I would recommend you following the instructions provided in the hoffman2 [website](https://www.hoffman2.idre.ucla.edu/Using-H2/Connecting/Connecting.html).

# To Install R
Open an internet browser and go to www.r-project.org.
Click the "download R" link in the middle of the page under "Getting Started."
Select a CRAN location (a mirror site) and click the corresponding link.
Click on the "Download R for (Mac) OS X" link at the top of the page.
Click on the file containing the latest version of R under "Files."
Save the .pkg file, double-click it to open, and follow the installation instructions.
Now that R is installed, you need to download and install RStudio.
# To Install RStudio
Go to www.rstudio.com and click on the "Download RStudio" button.
Click on "Download RStudio Desktop."
Click on the version recommended for your system, or the latest Mac version, save the .dmg file on your computer, double-click it to open, and then drag and drop it to your applications folder.

Once we are all able to access the cluster we can then install [PLINK](https://www.cog-genomics.org/plink2/)!

First download the file from their [website](https://www.cog-genomics.org/plink2/): 
then go to the terminal in your computer and go to where your file is via *cd*, mine is in the Downloads folder, so I can type:

```{bash, eval=F}
cd Downloads
scp plink_linux_x86_64_20210606.zip mica20@hoffman2.idre.ucla.edu:/u/home/m/mica20/project-collaboratory/PLINK_exercises/
```

Now go to the open window you have that gives you access to the cluster and unzip the file in the cluster by typing:

```{bash, eval=F}
unzip plink_linux_x86_64_20210606.zip
```

Now you are ready to use plink, type the following to see if works
```{bash, eval=F}
./plink
```

# Converting VCF format to PLINK format

# Downloading data
Instead of using the vcf file let's just use

https://www.cog-genomics.org/plink/1.9/resources


# Quality Control Practical
We will now work with a set of data files containing many SNPs from chromosome 19 genotyped on controls and cases. Data from a GWAS would contain SNPs at this density across the entire genome, but we will focus on just one chromosome to make the exercises more tractable.

The key files are:
chr19-example.vcf
chr19-example.samples

As we saw above, we can convert these to files PLINK can read. Here, instead of converting to .ped files with “--recode", we will convert them to so-called “binary PLINK” or .bed format using “--make-bed”:

plink --make-bed --vcf chr19-example.vcf --pheno chr19-example.samples --update-sex chr19-example.samples  2 --out chr19-example --keep-allele-order

Like the .ped and .map above, these files contain information about the samples and SNPs, as well as the genotypes for each of the samples at each of the SNPs. Unlike the human-readable text .ped file we used before, these data are in “binary ped” format (.bed). This format is a compressed format which saves space and speeds up analysis. Information on samples can be found in the .fam file and information on SNPs in the .bim file. We've used the --keep-allele-order option here to make sure plink preserves the two alleles at each SNP in the same order as in the VCF file (we'll see why this is important later) 

We can load data in the binary plink format using the --bfile option. For instance, to calculate allele frequencies we use:

plink --bfile chr19-example --freq

Q. How many SNPs and samples are in this dataset?

Reformatting between .bed/.bim/.fam  and  .ped/.map is easy

plink --bfile chr19-example --recode --out chr19-example

Note that this may take a few moments to run. Look at the file sizes in the directory (ls -l [minus ell] on the terminal prompt). 

Q. What are the relative sizes of .ped, .bed and .vcf files?

Note: Many other formats are in use for genetic data, including 'GEN' format, 'BGEN' format, and BCF (binary VCF) format. 

There are many different QC metrics that we can calculate for our dataset. These metrics can tell us about the quality of loci (i.e. SNPs), and of samples. For instance, we can calculation information about missingness:

plink --bfile chr19-example --missing --out miss-info

This will produce a .imiss file with information about individuals and .lmiss with information about loci (SNPs). You can load this output into a spreadsheet program to look at it in more detail: In the directory browser right-click miss-info.lmiss file.  Select open in Libreoffice Calc or choose other application to select Libreoffice Calc.  An options window will open and in the Separator options you need to ensure that only 'separated by space' and the 'merge delimiters' boxes are checked before proceeding!

Q. Which SNP has the highest missing rate?

Another QC metric is sample heterozygosity:
plink --bfile chr19-example --het --out het-info

We can use R to visualize these results. Launch R as in previous practicals. Commands to be typed in R (rather than in the terminal) will be shaded like this:

In RStudio:
setwd('/media/ubuntu/data/GEIA/Practicals/07_GWAS_analysis/')
HetData<-read.table("het-info.het",h=T)
hist(HetData$F,breaks=100)

The graph you see shows the distribution of heterozygosity across samples. QC plots often look like this; there is a large peak of samples that lie within a range of normal values (the normal samples), and then a small number of outlier samples that are usually poor quality samples. In this case you can see that there is a large peak in heterozygosity around 0.1, with a number of outliers below 0 or above 0.2. 
You can save a picture of this analysis for later reference:

In RStudio:
png("HetHist.png")
hist(HetData$F,breaks=100)
dev.off()

Next, lets plot the missingness values.

In RStudio:
MissData<-read.table("miss-info.imiss",h=T)
hist(MissData$F_MISS,breaks=100)
png("MissHist.png")
hist(MissData$F_MISS,breaks=100)
dev.off()

Again, most of the samples have low missingness (close to zero), with a number of outliers above 0.02. We can now combine these two QC metrics (missingness and heterozygosity) to select outlying samples:
In RStudio:
qcFails <- MissData[MissData$F_MISS > 0.02 | HetData$F > 0.2 | HetData$F < 0,c(1,2)]
write.table(qcFails,quote=F,row.names=F,col.names=F,file="qcFails.txt")


We can combine this list with other quality control metrics (described in the lectures) to create a clean dataset (note that although we need to break long commands in this document over two lines, all options should be typed on one continuous line in the terminal without breaks):

plink --bfile chr19-example --remove qcFails.txt --hwe 1e-4 include-nonctrl --geno 0.01 --maf 0.01 --make-bed --out chr19-clean --keep-allele-order

Q. What filters have we applied here? (the PLINK website can help explain these options)


